### PR TITLE
Fixes to help preventing "There is no singleton instance errors"

### PR DESCRIPTION
**Pull request by vegaro**
Fixes [sc-11174]

### CODE REVIEW

File: src/purchases.ts [hunk 1]
```diff
@@ -25,8 +25,8 @@ const eventEmitter = new NativeEventEmitter(RNPurchases);
  * @param {Object} purchaserInfo Object containing info for the purchaser
  */
 export type PurchaserInfoUpdateListener = (purchaserInfo: PurchaserInfo) => void;
-export type ShouldPurchasePromoProductListener = (deferredPurchase: () => MakePurchasePromise) => void;
-type MakePurchasePromise = Promise<{ productIdentifier: string; purchaserInfo: PurchaserInfo; }>;
+export type ShouldPurchasePromoProductListener = (deferredPurchase: () => Promise<MakePurchaseResult>) => void;
+type MakePurchaseResult = { productIdentifier: string; purchaserInfo: PurchaserInfo; };
```
[the remark this answers is not in the document]
Comment (taquitos) [reply src/purchases.ts:29]: I feel like this is easier to understand, anyway

File: src/purchases.ts [hunk 2]
```diff
@@ -207,14 +207,16 @@ export default class Purchases {
      * the new ID with the previous id.
      * @param {boolean} allowSharing Set this to true if you are passing in an appUserID but it is anonymous, this is true by default if you didn't pass an appUserID
      */
-    public static setAllowSharingStoreAccount(allowSharing: boolean): void {
+    public static async setAllowSharingStoreAccount(allowSharing: boolean): Promise<void> {
```
[the remark this answers is not in the document]
Comment (taquitos) [reply src/purchases.ts:210]: With citations 😲

File: src/purchases.ts [hunk 3]
```diff
@@ -818,9 +859,26 @@ export default class Purchases {
...
+     * @returns {Promise<Boolean>} promise with boolean response
+     */
+    public static isConfigured(): Promise<boolean> {
+        return RNPurchases.isConfigured();
+    }
+
+    private static async throwIfNotConfigured() {
+        const isConfigured = await Purchases.isConfigured();
+        if (!isConfigured) {
+            throw new Error("There is no singleton instance. " +
```
[the remark this answers is not in the document]
Comment (taquitos) [reply src/purchases.ts:878]: I'm a fan of custom errors, but, what is the standard practice for RN?

File: __tests__/index.test.js [hunk 4]
```diff
@@ -775,4 +776,28 @@ describe("Purchases", () => {
         });
       });
   });
+
+  describe("isConfigured", () => {
```
Comment (alfondotnet) [inline __tests__/index.test.js:780]: shouldn't we also test that we throw when it's not configured?

Comment (alfondotnet) [inline __tests__/index.test.js:780]: Questions without a lot of context: isn't there any test we need to also add that iterates over all the appropriate methods and expects that we throw, if they are called before setup?

Comment (vegaro) [reply __tests__/index.test.js:780]: that would be a good one. I guess I can store all functions we care about in an array and iterate over it?

Comment (vegaro) [reply __tests__/index.test.js:780]: I added a test, I am not sure there's a better way of doing it but this works and will fail if a new function is added and it requires to check the instance [`6f3a8f0` (#310)](https://github.com/RevenueCat/react-native-purchases/pull/310/commits/6f3a8f04aa73c2d360b62cc57964b947b95ce2d7)

File: src/purchases.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (vegaro) [reply src/purchases.ts]: I created a new `UninitializedPurchasesError` 🤷‍♂️

File: ios/RNPurchases.m [hunk 5]
```diff
@@ -320,6 +320,12 @@ - (dispatch_queue_t)methodQueue {
       resolve(@([RCCommonFunctionality canMakePaymentsWithFeatures:features]));
 }
 
+RCT_REMAP_METHOD(isConfigured,
+                 isConfiguredWithResolve:(RCTPromiseResolveBlock)resolve
+                 reject:(RCTPromiseRejectBlock)reject) {
+    resolve(@([RCPurchases isConfigured]));
```
[the remark this answers is not in the document]
Comment (aboedo) [reply ios/RNPurchases.m:326]: `isConfigured` is a property, so we should use the dot notation to read it, i.e.: 
```suggestion
    resolve(@(RCPurchases.isConfigured));
```

File: src/purchases.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (aboedo) [reply src/purchases.ts:212]: so this was always wrong, correct?

File: dist/purchases.d.ts [hunk 6]
```diff
@@ -366,118 +367,118 @@ export default class Purchases {
...
         [key: string]: string | null;
-    }): void;
+    }): Promise<void>;
     /**
      * Subscriber attribute associated with the email address for the user
      *
      * @param email Empty String or null will delete the subscriber attribute.
      */
-    static setEmail(email: string | null): void;
+    static setEmail(email: string | null): Promise<void>;
```
Comment (beylmk) [inline dist/purchases.d.ts:376]: should we add a comment to all of the methods that now return a promise, to say that they'll throw?

File: ios/RNPurchases.m (hunk 5 above)
[the remark this answers is not in the document]
Comment (vegaro) [reply ios/RNPurchases.m:326]: Which reminds me I haven't tested this in iOS!?? I will do that ASAP, thanks for catching that

File: src/purchases.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (vegaro) [reply src/purchases.ts:212]: Basically. Everything is asynchronous, so yes. Having it returning `void` makes it look like it's sync, but we are not marking it with `isBlockingSynchronousMethod = true`.

**Review by aboedo — APPROVED** [review]

**Comment by taquitos** [comment]
FYI the test failure seems to be: https://discuss.circleci.com/t/unrecognized-vm-option-usecgroupmemorylimitforheap/37088

**Comment by aboedo** [comment]
@vegaro what's the status on this one?

**Comment by vegaro** [comment]
Looking at it the failure now so we can merge it @aboedo

**Comment by vegaro** [comment]
ok, looks like the more recent version of the orb has the right JVM option https://github.com/react-native-community/react-native-circleci-orb/blob/master/src/executors/linux_android.yml#L5

So upgrading!!

**Pull request merged**